Allow asymmetric matchers (jest) and partial matchers (jasmine)

sfargier opened the pull request:
This PR enables usage of asymmetric/partial matchers (e.g. expect.stringContaining and jasmine.stringMatching) with toHaveProp, toHaveState and toHaveStyle. This fixes #114.
Example:
```javascript
// jest (asymmetric matcher)
expect(wrapper).toHaveProp('prop1', expect.objectContaining({ simple1: 'value1' }));
// jasmine (partial matchers)
expect(wrapper).toHaveState('array', jasmine.arrayContaining([2]));
```

To sum it up, usage of deepEqualIdent is replaced with the equals utility method from the jest/jasmine when available (deepEqualIdent is kept as fallback just in case).

Notes:
- Jest updated to version 20 to be able to use the new this.equals method inside matchers
- In jasmine enzyme, the matchers exported as jasmineEnzyme.enzymeMatchers are unchanged (partial matchers won't work). If I understood it right, the 1.x versions of jasmine do not have the equals utility function available.
- Ideally, a few more matchers (for example toHaveHtml, toHaveText) are working with string and could be updated to enable the use of expect.stringContaining. But some of them are explicitly expecting string and manipulate them, which is not possible when receiving the asymmetric matcher. So I only updated the matcher whose parameter have an "any" type declaration.

sfargier commented:
Oops, I updated the jest dependency with npm instead of yarn so the yarn.lock didn't get updated.
Yarn install does not work locally at the moment because mirrors.gandi.net seems to be unreachable, so I'll try again later and fix the PR then.

blainekasten commented on packages/jasmine-enzyme/src/index.js:
```diff
@@ -21,11 +21,17 @@ function jasmineEnzyme(): void {
+  const toJasmineMatcher = (matcherFn: Function) => (util: Object, customEqualityTesters: Object) => {
+    // Convert the equals util from jasmine to share the same interface as jest
+    const equals = (actual, expected) => util.equals(actual, expected, customEqualityTesters);
+    return { compare: matcherFn.bind({ equals }) };
```
I like this solution! Nice work

sfargier commented:
Thanks.
I rebased and squashed the commits, so this should be good. I also redid the dependency update so that the yarn.lock file gets updated properly this time, if you wanna check it (I confirm yarn works fine for me too with your latest commit).

blainekasten commented:
Amazing work, thanks @sfargier
